util/src/main/java/io/kubernetes/client/util/ModelMapper.java

```diff
@@ -42,16 +44,23 @@
 
   public static final Duration DEFAULT_DISCOVERY_REFRESH_INTERVAL = Duration.ofMinutes(30);
 
-  private static Map<GroupVersionKind, Class<?>> classes = new HashMap<>();
-
-  private static Set<Discovery.APIResource> lastAPIDiscovery = new HashSet<>();
-
-  private static volatile long nextRefreshTimeMillis = System.currentTimeMillis();
+  // initially loaded from classpath, can be incorrect
+  private static Map<GroupVersionKind, Class<?>> preBuiltClassesByGVK = new HashMap<>();
 
   // Model's api-group prefix to kubernetes api-group
-  private static Map<String, String> apiGroups = new HashMap<>();
+  private static Map<String, String> preBuiltApiGroups = new HashMap<>();
+
   // Model's api-version midfix to kubernetes api-version
-  private static List<String> apiVersions = new ArrayList<>();
+  private static List<String> preBuiltApiVersions = new ArrayList<>();
+
+  private static Map<GroupVersionKind, Class<?>> classesByGVK = new ConcurrentHashMap<>();
+
+  private static Map<GroupVersionResource, Class<?>> classesByGVR = new ConcurrentHashMap<>();
+
+  private static Set<Discovery.APIResource> lastAPIDiscovery = new HashSet<>();
```

Comment:
Should this be a ConcurrentHashMap too? (you can use it like a `Set`)

Reply:
the map is actually thread-safe b/c it's immutable. we're replace the `lastAPIDiscovery` w/ a new immutable map upon every refreshing